modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliverableAndesMetadata.java

```diff
@@ -0,0 +1,544 @@
+    }
+
+    /**
+     * Mark as deleted message
+     */
+    public void markAsDeletedMessage() {
+        addMessageStatus(MessageStatus.DELETED);
+    }
+
+    /**
+     * Check if metadata is stale
+     *
+     * @return true if message is stale
+     */
+    public boolean isStale() {
+        return stale;
+    }
+
+    public void markAsStale() {
+        stale = true;
+    }
+
+    /**
+     * Mark as slot removed message
+     */
+    public void markAsSlotRemoved() {
+        addMessageStatus(MessageStatus.SLOT_REMOVED);
+    }
+
+    /**
+     * Due to last subscription close of local node
+     * slots can get returned to slot coordinator. There we mark
+     * the messages in that slot as slot returned
+     */
+    public void markAsSlotReturned() {
+        addMessageStatus(MessageStatus.SLOT_RETURNED);
+    }
+
+    /**
+     * Cancel message delivery for channel. This is called when a
+     * message delivery is failed from broker side. By the time message MUST be maked
+     * as SENT (we assume and mark SENT before actual send) to this channel
+     * @param channelID id of the channel
+     * @return current number of times this message is delivered to the given channel
+     */
+    public int markDeliveryFailureOfASentMessage(UUID channelID) {
+        channelDeliveryInfo.get(channelID).
+                addChannelStatus(ChannelMessageStatus.SEND_FAILED);
+        return channelDeliveryInfo.get(channelID).decrementDeliveryCount();
+    }
+
+    /**
+     * Remove a channel that this message is scheduled to deliver. If the subscriber has closed
+     * during the message schedule and actual sent this should be called
+     * @param channelID Id of the channel to remove
+     */
+    public void removeScheduledDeliveryChannel(UUID channelID) {
+        channelDeliveryInfo.remove(channelID);
+        if(isMarkAsDelivered()) {
+            addMessageStatus(MessageStatus.SENT_TO_ALL);
+        }
+        if(isMarkAsAcked()) {
+            addMessageStatus(MessageStatus.ACKED_BY_ALL);
+        }
+    }
+
+
+    /**
+     * Get the channels this message is delivered to
+     * @return Set of channel IDs
+     */
+    public Set<UUID> getAllDeliveredChannels() {
+        return channelDeliveryInfo.keySet();
+    }
+
+    /**
+     * Check if this message delivered to all the scheduled channels
+     * @return true if delivered to all the channels
+     */
+    public boolean isMarkAsDelivered() {
+        boolean isDelivered = true;
+        for (Map.Entry<UUID, ChannelInformation> channelInfoEntry : channelDeliveryInfo.entrySet()) {
+            ChannelMessageStatus channelMessageStatus = channelInfoEntry.getValue().getLatestMessageStatus();
+            if(null == channelMessageStatus
+                    || !(channelMessageStatus.equals(ChannelMessageStatus.SENT)
+                    || channelMessageStatus.equals(ChannelMessageStatus.RESENT)
+                    || channelMessageStatus.equals(ChannelMessageStatus.ACKED))) {
```

Comment:
Won't this condition check lead to an NPE? IMO this is ok if there are only &&. Here we have ||